# Sentiment Analysis Using TF-IDF (Term Frequency-Inverse Document Frequency) and Logistic Regression

## Project Overview

Natural Language Processing (NLP) enables machines to understand, analyze, and extract meaningful information from human language. One of the most fundamental NLP tasks is **Sentiment Analysis**, which aims to determine whether a piece of text expresses a positive or negative opinion.

In this project, I build a complete sentiment analysis pipeline using classical NLP techniques and Machine Learning algorithms. The workflow covers the entire process from raw text preprocessing to feature extraction, model training, evaluation, and experiment documentation.

The primary objective of this project is not only to achieve high predictive performance but also to systematically investigate how different preprocessing techniques, feature engineering strategies, and machine learning models affect sentiment classification performance.

---

## Dataset

Dataset Link: **[IMBD Movie Reviews](https://www.kaggle.com/datasets/mwallerphunware/imbd-movie-reviews-for-binary-sentiment-analysis)**

The dataset consists of movie reviews labeled with their corresponding sentiment:

* Positive
* Negative

The reviews contain real-world natural language, making them suitable for evaluating text preprocessing techniques and machine learning models.

---

## Project Pipeline

The following stages are implemented throughout this project:

### 1. Text Preprocessing

* Contraction expansion
* Lowercasing
* Text cleaning using regular expressions
* Stopword removal with preserved negations
* Lemmatization
* Corpus construction

### 2. Feature Extraction

* Term Frequency-Inverse Document Frequency (TfidfVectorizer)
* N-gram generation
* Vocabulary analysis
* Feature selection using `max_features`
* Rare-word filtering using `min_df`

### 3. Model Training

Different machine learning algorithms are evaluated and compared, including:

* Gaussian Naive Bayes
* Multinomial Naive Bayes
* Bernoulli Naive Bayes
* Logistic Regression
* Support Vector Machines (SVM)
* Decision Trees
* Random Forests

### 4. Model Evaluation

Performance is assessed using multiple metrics:

* Accuracy
* Precision
* Recall
* F1-Score
* ROC-AUC Score
* Confusion Matrix
* Cross-Validation Mean Accuracy
* Cross-Validation Standard Deviation

---

## Experimental Approach

Rather than training a single model, this notebook follows an experimentation-driven methodology.

For each model, multiple configurations are tested, including different values for:

* `max_features`
* `min_df`
* N-gram ranges
* Preprocessing strategies

The goal is to identify the most effective configuration while understanding the trade-offs between model complexity, computational cost, and predictive performance.

---

## Key Learning Objectives

Through this project, I aim to:

* Develop a deeper understanding of NLP preprocessing techniques.
* Compare the behavior of different machine learning algorithms on text data.
* Analyze how feature engineering impacts classification performance.
* Build reproducible NLP pipelines suitable for real-world applications.
* Establish strong baselines before moving toward advanced embedding and transformer-based approaches.

---

**Author:** Hazem Mohamed

**Role:** AI Engineer | Machine Learning Engineer | NLP Engineer

**Repository:** [NLP Experimentation Lab](https://github.com/Hazem1695/NLP-Experimentation-Lab)


# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

## **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv('MovieReviewTrainingDatabase.csv')
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  25000 non-null  object
 1   review     25000 non-null  object
dtypes: object(2)
memory usage: 390.8+ KB
None

SHAPE OF DATA
Rows: 25000, Columns: 2

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 96

DATA TYPES
sentiment    object
review       object
dtype: object

NUMERICAL SUMMARY
       sentiment                                             review
count      25000                                              25000
unique         2                                              24904
top     Positive  You do realize that you've been watching the E...
freq       12500                                                  3

CATEGORICAL SUMMARY
       sentiment                                             review
count      25000                 

## Duplicate Data Detection

In [4]:
duplicates = dataset[dataset.duplicated(subset=['review'], keep=False)]
duplicates.sort_values('review')

,sentiment,review
21186,Negative,"Back in his youth, the old man had wanted to..."
21877,Negative,"Back in his youth, the old man had wanted to..."
14734,Negative,'Dead Letter Office' is a low-budget film abou...
5519,Negative,'Dead Letter Office' is a low-budget film abou...
7011,Positive,".......Playing Kaddiddlehopper, Col San Fernan..."
...,...,...
2685,Negative,"in this movie, joe pesci slams dunks a basketb..."
22244,Positive,it's amazing that so many people that i know h...
14767,Positive,it's amazing that so many people that i know h...
12462,Negative,this movie begins with an ordinary funeral... ...


## Quantifying Duplicate Review Frequencies

In [5]:
review_counts = dataset['review'].value_counts()
print("Reviews appearing more than once:")
print((review_counts > 1).sum())
print("\nMaximum repetitions:")
print(review_counts.max())

Reviews appearing more than once:
92

Maximum repetitions:
3


## Removing Duplicate Reviews & Resetting Index
> **Note:** This cell drops the repeated rows we identified in the previous steps and cleanly resets the row indices for model training

In [6]:
print("Before:", len(dataset))
dataset = dataset.drop_duplicates()
print("After:", len(dataset))
dataset = dataset.reset_index(drop=True)

Before: 25000
After: 24904


## Library Installation
> **Note:** The `contractions` library is required to automatically expand shortcuts like *don't* to *do not* and *I'm* to *I am* during preprocessing.

In [7]:
!pip install contractions

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.1/345.1 kB 9.3 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 6.4 MB/s eta 0:00:00


## **Cleaning the texts**

In [8]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')
import re
import contractions
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

base_stopwords = set(stopwords.words('english'))
negation_words = {'not', 'no', 'never'}
all_stopwords = base_stopwords - negation_words

wnl = WordNetLemmatizer()

corpus = []

for i in range(0, len(dataset)):
    review = dataset['review'][i]
    # Fix contractions
    review = contractions.fix(review)
    # Lowercase
    review = review.lower()
    
    review = re.sub(r'[^a-zA-Z\s]', ' ', review)
    # Split
    words = review.split()
    # Chained Lemmatization (Handles both Verbs 'v' and Nouns 'n')
    review = [wnl.lemmatize(wnl.lemmatize(word, pos='v'), pos='n') for word in words if word not in all_stopwords]
    review = ' '.join(review) 
    corpus.append(review)

[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /usr/share/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /usr/share/nltk_data...


## Preprocessing Verification
> **Note:** Pulling the first two rows directly as a memory array to confirm that our lowercasing, stopword stripping, and lemmatization pipeline worked correctly before feeding it into the vectorizer.

In [9]:
# Pull the data directly as a fast memory array
raw_samples = dataset['review'].head(2).values

for i in range(2):
    print(f"=== REVIEW #{i+1} ===")
    print(f"RAW:     {raw_samples[i]}\n") 
    print(f"CLEANED: {corpus[i]}")
    print("-" * 50)

=== REVIEW #1 ===
RAW:     With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.  Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.  The actual feature film bit when it final

# **Encoding Categorical data Using Label Encoding**

In [10]:
from sklearn.preprocessing import LabelEncoder
y = dataset.iloc[:, 0].values
le = LabelEncoder()
y = le.fit_transform(y)

In [11]:
print(y)

[1 1 0 ... 0 0 1]


# **Splitting the dataset into the Training set and Test set**

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(corpus, y, test_size = 0.20, random_state = 0)

# Class Balance Check
> **Note:** Using NumPy to verify if our dataset is perfectly balanced between positive and negative reviews before splitting it into training and testing sets.

In [13]:
# This returns the unique classes and how many times they appear
classes, counts = np.unique(y, return_counts=True)
for c, count in zip(classes, counts):
    print(f"Class {c} contains {count}")

Class 0 contains 12432
Class 1 contains 12472


# **Creating the TF-IDF (Term Frequency-Inverse Document Frequency) model**

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Add ngram_range=(1, 2) so it automatically catches phrases like "not good" or "no clue"
tf = TfidfVectorizer(max_features=20000, min_df=2, ngram_range=(1, 2))
X_train = tf.fit_transform(X_train)
X_test = tf.transform(X_test)

In [15]:
print(X_train.shape)        # (n_samples, n_features)
print(X_train.shape[0])     # number of training samples
print(X_train.shape[1])     # number of features (vocabulary size)

(19923, 20000)
19923
20000


In [16]:
print(X_test.shape)        # (n_samples, n_features)
print(X_test.shape[0])     # number of training samples
print(X_test.shape[1])     # number of features (vocabulary size)

(4981, 20000)
4981
20000


# **Training the Logistic Regression model on the Training set**

In [17]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train,y_train)

LogisticRegression(random_state=0)

# **Predicting the Test set results**

In [18]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))# [0:100,:]

[[1 1]
 [1 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


# **Evaluating the Model Performance**

In [19]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=3)

print("\nMean Accuracy:")
print(accuracies.mean())

print("\nStandard Deviation:")
print(accuracies.std())

Confusion Matrix:
[[2202  318]
 [ 207 2254]]

Accuracy Score:
0.8945994780164626

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.87      0.89      2520
           1       0.88      0.92      0.90      2461

    accuracy                           0.89      4981
   macro avg       0.90      0.89      0.89      4981
weighted avg       0.90      0.89      0.89      4981


ROC-AUC Score:
0.8948486871384067

Mean Accuracy:
0.8858103699242083

Standard Deviation:
0.0014668111425189337


# Logistic Regression with TF-IDF – Experimental Report

---

# Experiment Overview

This experiment evaluates the performance of **Logistic Regression** combined with different **TF-IDF vectorization strategies** for binary text classification.

The primary objective is to investigate how various TF-IDF configurations—including vocabulary size, n-gram features, minimum document frequency, and Logistic Regression regularization—affect overall classification performance, model stability, and generalization.

All experiments were conducted using identical train/test splits to ensure fair and reproducible comparisons.

---

# Experimental Configurations

The following TF-IDF configurations were evaluated:

| Experiment   | TF-IDF Configuration                             | Logistic Regression  |
| ------------ | ------------------------------------------------ | -------------------- |
| Experiment 1 | Default TF-IDF                                   | Default Parameters   |
| Experiment 2 | max_features=5,000, ngram_range=(1,2)            | Default Parameters   |
| Experiment 3 | max_features=10,000, ngram_range=(1,2)           | Default Parameters   |
| Experiment 4 | max_features=15,000, ngram_range=(1,2)           | Default Parameters   |
| Experiment 5 | max_features=20,000, min_df=2, ngram_range=(1,2) | Default Parameters   |
| Experiment 6 | max_features=25,000, min_df=2, ngram_range=(1,2) | Default Parameters   |
| Experiment 7 | Same as Experiment 6 + C=0.5                     | Tuned Regularization |

---

# Performance Summary

| Configuration           |   Accuracy |    ROC-AUC | CV Mean Accuracy |      CV Std |
| ----------------------- | ---------: | ---------: | ---------------: | ----------: |
| Default TF-IDF          | **88.82%** |     0.8884 |           88.10% |     0.00072 |
| 5k Features             |     88.56% |     0.8858 |           88.12% |     0.00105 |
| 10k Features            |     88.94% |     0.8896 |           88.58% | **0.00061** |
| 15k Features            | **89.40%** |     0.8942 |       **88.68%** |     0.00166 |
| 20k Features + min_df=2 | **89.46%** | **0.8948** |           88.58% |     0.00147 |
| 25k Features + min_df=2 |     89.42% |     0.8944 |           88.63% |     0.00193 |
| 25k Features + C=0.5    |     88.86% |     0.8889 |           87.97% | **0.00038** |

---

# Best Performing Model

## TF-IDF

* max_features = 20,000
* min_df = 2
* ngram_range = (1,2)

## Logistic Regression

* Default Parameters
* random_state = 0

### Performance

| Metric              |      Score |
| ------------------- | ---------: |
| Accuracy            | **89.46%** |
| ROC-AUC             | **0.8948** |
| Precision (Class 0) |       0.91 |
| Recall (Class 0)    |       0.87 |
| Precision (Class 1) |       0.88 |
| Recall (Class 1)    |       0.92 |
| Macro F1-score      |   **0.89** |

Confusion Matrix

```
[[2202 318]
 [207 2254]]
```

The model correctly classified **4,456 out of 4,981 samples**, producing only **525 misclassifications**, demonstrating strong discrimination between the two sentiment classes.

---

# Detailed Experimental Analysis

## 1. Default TF-IDF

The baseline model achieved an accuracy of **88.82%**, demonstrating that Logistic Regression is already a strong linear classifier for sparse TF-IDF features.

The model produced balanced precision and recall across both classes, providing an excellent starting point for further optimization.

---

## 2. Increasing Vocabulary Size

Increasing the TF-IDF vocabulary from the default configuration to **10,000**, **15,000**, **20,000**, and **25,000** features consistently improved model performance.

Larger vocabularies captured more informative words and phrases, allowing the classifier to better distinguish subtle linguistic patterns.

Performance improved steadily until approximately **20,000 features**, after which the improvements became marginal.

---

## 3. Effect of Bi-grams

Using

```
ngram_range=(1,2)
```

enabled the model to capture both individual words and meaningful word pairs.

This richer feature representation improved semantic understanding compared with unigram-only features, leading to higher classification accuracy.

---

## 4. Effect of min_df=2

Introducing

```
min_df=2
```

removed extremely rare words that appeared only once in the corpus.

Benefits included:

* Reduced feature sparsity
* Lower noise
* Improved generalization
* Better computational efficiency

This preprocessing step contributed to the highest overall ROC-AUC score.

---

## 5. Regularization Experiment

The final experiment reduced the Logistic Regression parameter

```
C = 0.5
```

which increases regularization strength.

Although stronger regularization generally reduces overfitting, it also slightly limited the model's ability to learn complex decision boundaries.

Compared with the default regularization:

* Accuracy decreased
* ROC-AUC decreased
* Cross-validation mean decreased

This indicates that the default regularization provided a better balance between bias and variance for this dataset.

---

# Cross-Validation Analysis

Cross-validation results were remarkably consistent across all experiments.

Standard deviations ranged from only

* **0.00038**
* **0.00193**

Such extremely small variations indicate:

* Excellent reproducibility
* High model stability
* Strong robustness
* Minimal sensitivity to different training folds

This consistency suggests that Logistic Regression generalizes reliably across unseen data.

---

# Model Strengths

* Fast training and inference
* Excellent scalability for large text datasets
* Highly interpretable linear decision boundaries
* Strong performance on sparse TF-IDF vectors
* Stable cross-validation performance
* Low computational cost
* Minimal risk of overfitting
* Competitive ROC-AUC performance

---

# Observed Limitations

Although the model performs strongly, several limitations remain:

* Linear decision boundaries cannot capture highly complex language relationships.
* Semantic understanding is limited compared to neural language models.
* Performance gains diminish as vocabulary size increases beyond approximately 20,000 features.
* The model depends entirely on manually engineered TF-IDF representations rather than contextual embeddings.

---

# Key Findings

* Logistic Regression remains one of the strongest traditional machine learning algorithms for text classification.
* TF-IDF vocabulary size significantly influences performance.
* Bi-gram features consistently improve classification quality.
* Removing rare words with `min_df=2` enhances robustness.
* Increasing vocabulary beyond 20,000 features yields diminishing returns.
* Stronger regularization (`C=0.5`) slightly reduced predictive performance.
* Cross-validation demonstrates exceptional model stability across all experiments.

---

# Final Conclusion

The experiments demonstrate that **Logistic Regression combined with TF-IDF is an exceptionally strong baseline for sentiment classification**. Systematic feature engineering substantially improved predictive performance over the baseline configuration.

The optimal configuration employed **20,000 TF-IDF features**, **bi-grams**, and **minimum document frequency filtering**, achieving an **accuracy of 89.46%** and a **ROC-AUC score of 0.8948**.

Overall, the model delivers an excellent balance of predictive accuracy, computational efficiency, robustness, and interpretability. These characteristics make Logistic Regression with TF-IDF a reliable solution for large-scale text classification tasks and an excellent benchmark against which more advanced deep learning and transformer-based approaches can be evaluated.
